### PR TITLE
Add SelectAsArray, DangerousCreateFromUnderlyingArray

### PR DESCRIPTION
**Pull request by tmat**
Adds an overload of `SelectAsArray` that takes a predicate.
Adds `DangerousCreateFromUnderlyingArray`.
Adds `SerializableBytes.ToImmutableArray`.

These helpers are used by https://github.com/dotnet/roslyn/pull/34027

### CODE REVIEW

**Comment by tmat** [comment]
@agocke @heejaechang  PTAL

File: src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs [hunk 1]
```diff
@@ -244,6 +245,12 @@ public byte[] ToArray()
                 return array;
             }
 
+            public ImmutableArray<byte> ToImmutableArray()
+            {
+                var array = ToArray();
+                return ImmutableArrayExtensions.DangerousCreateFromUnderlyingArray(ref array);
```
Comment (agocke) [inline src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: Do we really need to do this? Can't we do the same thing by creating a builder and filling it?

At least with the Span API we're eventually going to remove it, but I don't think there is an equivalent API on ImmutableArray, is there?

Comment (tmat) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: I didn't want to rewrite to much of the SerializableBytes implementation. Feel free to make it better. I think having the `DangerousCreateFromUnderlyingArray` API is good - it is useful every now and then.

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 2]
```diff
@@ -513,6 +541,13 @@ internal static T[] DangerousGetUnderlyingArray<T>(this ImmutableArray<T> array)
         internal static ReadOnlySpan<T> AsSpan<T>(this ImmutableArray<T> array)
             => array.DangerousGetUnderlyingArray();
 
+        internal static ImmutableArray<T> DangerousCreateFromUnderlyingArray<T>(ref T[] array)
```
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: This seems terrifying.

Comment (tmat) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: Hence the name

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs [hunk 3]
```diff
@@ -513,6 +541,13 @@ internal static T[] DangerousGetUnderlyingArray<T>(this ImmutableArray<T> array)
...
+        {
+            var proxy = new ImmutableArrayProxy<T> { MutableArray = array };
+            array = null;
```
Comment (sharwell) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: :thought_balloon: I would normally use `Interlocked.Exchange` here, but if you're in a place where that matters for the code then there are bigger problems.

File: src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs (hunk 1 above)
Comment (agocke) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: I don't think convenience is a good enough reason to add an API this dangerous.

Comment (tmat) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: We already have the dangeous API here that goes the other direction. I don't see the point of having the other one elsewhere.

Comment (agocke) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: That method has a proposal to remove it when new APIs are available from ImmutableArray. I don't see a proposal to remove this API ever.

Comment (tmat) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: I'll file an issue if that helps.

Comment (agocke) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: An issue to do what? Remove it when the API is available on ImmutableArray itself?

Comment (tmat) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: Issue to update SerializableBytes to not use the API.

Comment (AlekseyTs) [inline src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: >return ImmutableArrayExtensions.DangerousCreateFromUnderlyingArray(ref array) [](start = 16, length = 77)

Why not use ```ImmutableArray<T>.Builder```. Create a builder of the right size ```ImmutableArray.CreateBuilder<T>(size)```, fill it with data and call ```MoveToImmutable```.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs (hunk 1 above)
Comment (agocke) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: I don't think we should file issues to fix mistakes that we can easily not make in the first place.

Comment (tmat) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: Because I'd need to rewrite all the code that works with the mutable arrays.

Comment (tmat) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: I have to draw a line somewhere. I can't update all code to use immutable arrays. I have other things to do.

File: src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: >internal [](start = 8, length = 8)

The fact that this is `internal` makes it much less terrifying to me.

Comment (tmat) [reply src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs]: Yeah, I would certainly not want this to be public API.

File: src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Utilities/SerializableBytes.cs]: Looked into this again and I believe using `DangerousCreateFromUnderlyingArray` is actually the correct approach here, not just a workaround. The `SerializableBytes` type implements a custom `Stream`. `Stream` APIs work with `byte[]`. We could duplicate the implementation of the [helper](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Utilities/SerializableBytes.cs,212) that's populating they `byte[]` to work with ImmutableArray builder but I don't think it's desirable to have a copy of the code. The sole purpose of `DangerousCreateFromUnderlyingArray` is to be a bridge between mutable and immutable array APIs, which is exactly how it's used here.

**Comment by tmat** [comment]
@agocke @AlekseyTs any other feedback?

**Comment by AlekseyTs** [comment]
I am fine with the change if @agocke signs-off

**Comment by agocke** [comment]
I don't foresee a lot of good use-cases in the compiler, only risk. I'd put it in workspaces, since that's where it's used.

**Comment by tmat** [comment]
> I'd put it in workspaces, since that's where it's used. 

I can do that but it doesn't make sense to me to have it in a different place than the other Dangerous API.

**Comment by agocke** [comment]
I'm going to remove that API as soon as we Span on ImmutableArray, so that won't be a problem soon.

**Comment by agocke** [comment]
That's fine. Let's file an issue and when we get the Span APIs from ImmutableArray I'll move the code into the workspace layer instead of deleting it.

**Comment by tmat** [comment]
Sounds like a plan. Thanks!

**Comment by tmat** [comment]
Filed https://github.com/dotnet/roslyn/issues/34870

**Pull request merged**